File: .submodules/nostos/src/nostos/pxe/serve.py
"""PXE serving — orchestrates python HTTP + dnsmasq subprocesses.

Designed for macOS + Linux operator hosts. dnsmasq requires sudo to bind
DHCP port 67. HTTP is user-level on port 9080.
"""

from __future__ import annotations

import atexit
import logging
import os
import shutil
import signal
import subprocess
import sys
import time
from dataclasses import dataclass
from pathlib import Path

from ..paths import Paths

log = logging.getLogger(__name__)

DEFAULT_HTTP_PORT = 9080
DEFAULT_DHCP_RANGE_START = "192.168.68.200"
DEFAULT_DHCP_RANGE_END = "192.168.68.210"
DEFAULT_TFTP_ROOT = "/tmp/nostos-tftp"


class ServeError(RuntimeError):
    """Raised on serve-time setup problems."""


@dataclass
class NetworkInfo:
    """Detected host networking."""

    interface: str
    ip: str


def detect_network(subnet_hint: str = "192.168.68.") -> NetworkInfo:
    """Find the ethernet interface carrying the target subnet.

    Replaces the old ``detect-mac-ip.sh``. Skips Wi-Fi when possible —
    Wi-Fi loses the DHCP race against most consumer routers.
    """
    import ipaddress

    try:
        # macOS + Linux both have `ifconfig` or `ip -o addr` — use stdlib socket
        import psutil  # type: ignore[import-not-found]
    except ImportError:
        return _detect_via_ifconfig(subnet_hint)

    # Prefer psutil when available (more reliable).
    addrs = psutil.net_if_addrs()  # type: ignore[attr-defined]
    for iface, family_list in addrs.items():
        if iface.startswith(("lo", "utun", "awdl", "llw", "bridge", "stf", "gif", "anpi", "ap")):
            continue
        for entry in family_list:
            if entry.family == 2:  # AF_INET
                try:
                    ip_obj = ipaddress.ip_address(entry.address)
                except ValueError:
                    continue
                if str(ip_obj).startswith(subnet_hint):
                    return NetworkInfo(interface=iface, ip=str(ip_obj))
    raise ServeError(
        f"no interface found carrying {subnet_hint}x/24. "
        "Connect the operator host to the same switch as the target node."
    )


def _detect_via_ifconfig(subnet_hint: str) -> NetworkInfo:
    """Fallback: parse `ifconfig` output."""
    import re

    try:
        result = subprocess.run(
            ["ifconfig"], check=True, capture_output=True, text=True, timeout=5
        )
    except (subprocess.CalledProcessError, FileNotFoundError) as e:
        raise ServeError("cannot run ifconfig; install psutil or pass --iface") from e
    current_iface = None
    for line in result.stdout.splitlines():
        m = re.match(r"^(\w+):", line)
        if m:
            current_iface = m.group(1)
            continue
        if current_iface and subnet_hint in line:
            m2 = re.search(rf"inet ({re.escape(subnet_hint)}\d+)", line)
            if m2:
                if current_iface.startswith(("lo", "utun", "awdl", "llw", "bridge", "stf", "gif", "anpi", "ap")):
                    continue
                return NetworkInfo(interface=current_iface, ip=m2.group(1))
    raise ServeError(f"no interface found carrying {subnet_hint}x/24")


class Server:
    """Run HTTP + dnsmasq subprocesses in the foreground until interrupted."""

    def __init__(
        self,
        paths: Paths,
        *,
        http_port: int = DEFAULT_HTTP_PORT,
        dhcp_range_start: str = DEFAULT_DHCP_RANGE_START,
        dhcp_range_end: str = DEFAULT_DHCP_RANGE_END,
        gateway: str = "192.168.68.1",
        tftp_root: str = DEFAULT_TFTP_ROOT,
        iface: str | None = None,
    ) -> None:
        self.paths = paths
        self.http_port = http_port
        self.dhcp_range = (dhcp_range_start, dhcp_range_end)
        self.gateway = gateway
        self.tftp_root = Path(tftp_root)
        self.iface = iface  # None → auto-detect
        self.http_proc: subprocess.Popen[bytes] | None = None
        self.dnsmasq_proc: subprocess.Popen[bytes] | None = None

    def preflight(self) -> NetworkInfo:
        """Verify assets exist, tools are available, detect interface."""
        for required in ("ipxe.efi", "boot.ipxe"):
            if not (self.paths.assets / required).is_file():
                raise ServeError(
                    f"missing {required} under {self.paths.assets}. Run `nostos build` first."
                )
        if shutil.which("dnsmasq") is None and not Path("/opt/homebrew/sbin/dnsmasq").is_file():
            raise ServeError(
                "dnsmasq not found. Install via: brew install dnsmasq"
            )
        net = detect_network() if self.iface is None else NetworkInfo(interface=self.iface, ip="unknown")
        log.info("serving on %s (%s)", net.interface, net.ip)
        return net

    def stage_tftp(self) -> None:
        """Copy ipxe.efi to a world-readable path dnsmasq/nobody can read."""
        self.tftp_root.mkdir(parents=True, exist_ok=True)
        src = self.paths.assets / "ipxe.efi"
        dst = self.tftp_root / "ipxe.efi"
        shutil.copyfile(src, dst)
        os.chmod(self.tftp_root, 0o755)
        os.chmod(dst, 0o644)

    def kill_stale_http(self) -> None:
        """Kill any stale process bound to the configured HTTP port."""
        try:
            out = subprocess.run(
                ["lsof", "-ti", f":{self.http_port}"],
                capture_output=True,
                text=True,
                timeout=5,
            )
        except (subprocess.TimeoutExpired, FileNotFoundError):
            return
        for pid_str in out.stdout.strip().splitlines():
            try:
                pid = int(pid_str)
            except ValueError:
                continue
            log.info("killing stale process on :%s (pid %s)", self.http_port, pid)
            try:
                os.kill(pid, signal.SIGTERM)
            except OSError:
                pass
        if out.stdout.strip():
            time.sleep(1)

    def start_http(self) -> None:
        self.http_proc = subprocess.Popen(
            [sys.executable, "-m", "http.server", str(self.http_port)],
            cwd=str(self.paths.assets),
        )
        log.info("HTTP server PID %s on :%s", self.http_proc.pid, self.http_port)

    def start_dnsmasq(self, net: NetworkInfo) -> None:
        dnsmasq_bin = "/opt/homebrew/sbin/dnsmasq"
        if not Path(dnsmasq_bin).is_file():
            dnsmasq_bin = shutil.which("dnsmasq") or "dnsmasq"

        args: list[str] = [
            "sudo",
            dnsmasq_bin,
            "--no-daemon",
            "--port=0",
            f"--interface={net.interface}",
            f"--dhcp-range={self.dhcp_range[0]},{self.dhcp_range[1]},255.255.255.0,5m",
            f"--dhcp-option=3,{self.gateway}",
            f"--dhcp-option=6,{self.gateway}",
            "--dhcp-authoritative",
            "--dhcp-match=set:pxe,60,PXEClient",
            "--dhcp-ignore=tag:!pxe",
            "--enable-tftp",
            f"--tftp-root={self.tftp_root}",
            "--dhcp-userclass=set:ipxe,iPXE",
            f"--dhcp-boot=tag:!ipxe,ipxe.efi,,{net.ip}",
            f"--dhcp-boot=tag:ipxe,http://{net.ip}:{self.http_port}/boot.ipxe",
            "--log-queries",
            "--log-dhcp",
        ]
        self.dnsmasq_proc = subprocess.Popen(args)

    def stop(self) -> None:
        """Terminate children on shutdown."""
        for proc, name in [
            (self.dnsmasq_proc, "dnsmasq"),
            (self.http_proc, "http"),
        ]:
            if proc is None:
                continue
            if proc.poll() is None:
                log.info("stopping %s (pid %s)", name, proc.pid)
                try:
                    proc.terminate()
                    proc.wait(timeout=5)
                except subprocess.TimeoutExpired:
                    proc.kill()
                except OSError:
                    pass

    def run(self) -> int:
        """Foreground runloop. Returns dnsmasq's exit code."""
        net = self.preflight()
        self.kill_stale_http()
        self.stage_tftp()
        self.start_http()

        atexit.register(self.stop)

        def _sig_handler(signum: int, frame) -> None:
            self.stop()
            sys.exit(128 + signum)

        signal.signal(signal.SIGINT, _sig_handler)
        signal.signal(signal.SIGTERM, _sig_handler)

        self.start_dnsmasq(net)
        assert self.dnsmasq_proc is not None
        try:
            return self.dnsmasq_proc.wait()
        finally:
            self.stop()


def tear_down_stale(http_port: int = DEFAULT_HTTP_PORT) -> None:
    """`nostos serve --down` implementation."""
    try:
        subprocess.run(
            ["pkill", "-f", "python3 -m http.server " + str(http_port)],
            check=False,
            timeout=5,
        )
        subprocess.run(
            ["sudo", "-n", "pkill", "-f", "dnsmasq.*tftp-root"],
            check=False,
            timeout=5,
        )
    except (subprocess.TimeoutExpired, FileNotFoundError):
        pass

File: .submodules/nostos/internal/pxe/serve.go
package pxe

import (
	"context"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/yurifrl/nostos/internal/paths"
)

const (
	DefaultHTTPPort        = 9080
	DefaultDHCPRangeStart  = "192.168.68.200"
	DefaultDHCPRangeEnd    = "192.168.68.210"
	DefaultGateway         = "192.168.68.1"
	DefaultTFTPStaging     = "/tmp/nostos-tftp"
)

// NetworkInfo is the detected operator-host networking.
type NetworkInfo struct {
	Interface string
	IP        string
}

// Server supervises the HTTP + dnsmasq subprocesses for a PXE boot session.
type Server struct {
	Paths           paths.Paths
	HTTPPort        int
	Gateway         string
	DHCPRangeStart  string
	DHCPRangeEnd    string
	TFTPRoot        string
	Interface       string // override auto-detect
	ProxyMode       bool   // coexist with consumer's existing DHCP (recommended)

	httpSrv       *http.Server
	dnsmasqProc   *exec.Cmd
	httpRequests  chan string // relayed to consumer for progress tracking
	mu            sync.Mutex
}

// NewServer constructs a server with sensible defaults.
func NewServer(p paths.Paths) *Server {
	return &Server{
		Paths:          p,
		HTTPPort:       DefaultHTTPPort,
		Gateway:        DefaultGateway,
		DHCPRangeStart: DefaultDHCPRangeStart,
		DHCPRangeEnd:   DefaultDHCPRangeEnd,
		TFTPRoot:       DefaultTFTPStaging,
		ProxyMode:      true, // coexist with consumer's DHCP by default
	}
}

// HTTPRequests returns a channel of paths served (e.g. "/configs/...").
// Only populated while the server runs.
func (s *Server) HTTPRequests() <-chan string {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.httpRequests == nil {
		s.httpRequests = make(chan string, 64)
	}
	return s.httpRequests
}

// Preflight verifies assets + dnsmasq presence, detects interface, returns NetworkInfo.
func (s *Server) Preflight() (NetworkInfo, error) {
	for _, req := range []string{"ipxe.efi", "boot.ipxe"} {
		if _, err := os.Stat(filepath.Join(s.Paths.Assets(), req)); err != nil {
			return NetworkInfo{}, fmt.Errorf("missing %s under %s; run `nostos build` first", req, s.Paths.Assets())
		}
	}
	if !dnsmasqAvailable() {
		return NetworkInfo{}, errors.New("dnsmasq not found; install: brew install dnsmasq")
	}
	if s.Interface != "" {
		ip, err := ipForInterface(s.Interface)
		if err != nil {
			return NetworkInfo{}, fmt.Errorf("interface %s: %w", s.Interface, err)
		}
		return NetworkInfo{Interface: s.Interface, IP: ip}, nil
	}
	return detectNetwork()
}

// StageTFTP copies ipxe.efi into a world-readable staging path for dnsmasq.
func (s *Server) StageTFTP() error {
	if err := os.MkdirAll(s.TFTPRoot, 0o755); err != nil {
		return err
	}
	src := filepath.Join(s.Paths.Assets(), "ipxe.efi")
	dst := filepath.Join(s.TFTPRoot, "ipxe.efi")
	if err := copyFile(src, dst); err != nil {
		return err
	}
	return os.Chmod(dst, 0o644)
}

// KillStaleHTTP kills any process currently bound to the HTTP port.
func (s *Server) KillStaleHTTP() {
	out, err := exec.Command("lsof", "-ti", fmt.Sprintf(":%d", s.HTTPPort)).Output()
	if err != nil || len(out) == 0 {
		return
	}
	for _, line := range strings.Fields(string(out)) {
		var pid int
		fmt.Sscanf(line, "%d", &pid)
		if pid > 0 {
			slog.Info("killing stale process", "port", s.HTTPPort, "pid", pid)
			syscall.Kill(pid, syscall.SIGTERM)
		}
	}
	time.Sleep(500 * time.Millisecond)
}

// KillStaleDnsmasq asks sudo to kill any leftover nostos-managed dnsmasq.
func (s *Server) KillStaleDnsmasq() {
	_ = exec.Command("sudo", "-n", "pkill", "-f", "dnsmasq.*tftp-root="+s.TFTPRoot).Run()
	time.Sleep(500 * time.Millisecond)
}

// Start launches HTTP + dnsmasq. Returns on ctx cancel or dnsmasq exit.
// Progress events are surfaced via HTTPRequests().
func (s *Server) Start(ctx context.Context, net NetworkInfo) error {
	s.KillStaleHTTP()
	s.KillStaleDnsmasq()
	if err := s.StageTFTP(); err != nil {
		return fmt.Errorf("stage TFTP: %w", err)
	}
	if err := s.startHTTP(); err != nil {
		return fmt.Errorf("HTTP: %w", err)
	}
	if err := s.startDnsmasq(net); err != nil {
		s.Stop()
		return fmt.Errorf("dnsmasq: %w", err)
	}
	slog.Info("PXE server listening", "iface", net.Interface, "ip", net.IP, "port", s.HTTPPort)
	return nil
}

// Wait blocks until dnsmasq exits (or ctx is cancelled) and then stops everything.
func (s *Server) Wait(ctx context.Context) error {
	if s.dnsmasqProc == nil {
		return errors.New("dnsmasq not running")
	}
	done := make(chan error, 1)
	go func() {
		done <- s.dnsmasqProc.Wait()
	}()
	select {
	case <-ctx.Done():
		s.Stop()
		return ctx.Err()
	case err := <-done:
		s.Stop()
		return err
	}
}

// Stop terminates the subprocesses.
func (s *Server) Stop() {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.dnsmasqProc != nil && s.dnsmasqProc.Process != nil {
		_ = s.dnsmasqProc.Process.Signal(syscall.SIGTERM)
		// give sudo + dnsmasq a moment
		done := make(chan struct{})
		go func() { s.dnsmasqProc.Wait(); close(done) }()
		select {
		case <-done:
		case <-time.After(3 * time.Second):
			_ = s.dnsmasqProc.Process.Kill()
		}
	}
	if s.httpSrv != nil {
		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		_ = s.httpSrv.Shutdown(ctx)
		cancel()
	}
	if s.httpRequests != nil {
		close(s.httpRequests)
		s.httpRequests = nil
	}
}

// --- internals ---

func (s *Server) startHTTP() error {
	handler := loggingMiddleware(http.FileServer(http.Dir(s.Paths.State())), s)
	s.httpSrv = &http.Server{
		Addr:         fmt.Sprintf(":%d", s.HTTPPort),
		Handler:      handler,
		ReadTimeout:  30 * time.Second,
		WriteTimeout: 5 * time.Minute, // initramfs is 100+ MB
	}
	ln, err := net.Listen("tcp", s.httpSrv.Addr)
	if err != nil {
		return err
	}
	go func() {
		if err := s.httpSrv.Serve(ln); err != nil && err != http.ErrServerClosed {
			slog.Error("HTTP serve error", "err", err)
		}
	}()
	return nil
}

func (s *Server) startDnsmasq(ni NetworkInfo) error {
	bin := "/opt/homebrew/sbin/dnsmasq"
	if _, err := os.Stat(bin); err != nil {
		if path, e := exec.LookPath("dnsmasq"); e == nil {
			bin = path
		}
	}

	var dhcpRangeArg string
	if s.ProxyMode {
		// Proxy mode: we answer only the PXE/bootfile part. Deco (or whatever
		// existing DHCP server exists on the LAN) keeps giving out IPs.
		dhcpRangeArg = "--dhcp-range=" + s.Gateway + ",proxy"
	} else {
		dhcpRangeArg = fmt.Sprintf("--dhcp-range=%s,%s,255.255.255.0,5m", s.DHCPRangeStart, s.DHCPRangeEnd)
	}

	args := []string{
		bin,
		"--no-daemon",
		"--port=0",
		"--interface=" + ni.Interface,
		dhcpRangeArg,
		"--dhcp-match=set:pxe,60,PXEClient",
		"--dhcp-ignore=tag:!pxe",
		"--enable-tftp",
		"--tftp-root=" + s.TFTPRoot,
		"--dhcp-userclass=set:ipxe,iPXE",
		fmt.Sprintf("--dhcp-boot=tag:!ipxe,ipxe.efi,,%s", ni.IP),
		fmt.Sprintf("--dhcp-boot=tag:ipxe,http://%s:%d/assets/boot.ipxe", ni.IP, s.HTTPPort),
		"--pxe-prompt=nostos boot,1", // force PXE clients to pick our bootfile immediately
		"--log-queries",
		"--log-dhcp",
	}

	if !s.ProxyMode {
		// Full-DHCP mode: we need to provide gateway/router info too.
		args = append(args,
			"--dhcp-option=3,"+s.Gateway,
			"--dhcp-option=6,"+s.Gateway,
			"--dhcp-authoritative",
		)
	}

	cmd := exec.Command("sudo", args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	if err := cmd.Start(); err != nil {
		return err
	}
	s.dnsmasqProc = cmd
	return nil
}

// loggingMiddleware emits each served path to the HTTPRequests channel.
type srvLogger struct {
	s *Server
	w http.ResponseWriter
}

func (l *srvLogger) Header() http.Header         { return l.w.Header() }
func (l *srvLogger) Write(b []byte) (int, error) { return l.w.Write(b) }
func (l *srvLogger) WriteHeader(c int)            { l.w.WriteHeader(c) }

func loggingMiddleware(next http.Handler, s *Server) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		s.mu.Lock()
		ch := s.httpRequests
		s.mu.Unlock()
		if ch != nil {
			select {
			case ch <- r.URL.Path:
			default: // drop if consumer is behind
			}
		}
		next.ServeHTTP(w, r)
	})
}

// ipForInterface looks up the first IPv4 192.168.68.x address of an interface.
func ipForInterface(name string) (string, error) {
	iface, err := net.InterfaceByName(name)
	if err != nil {
		return "", err
	}
	addrs, err := iface.Addrs()
	if err != nil {
		return "", err
	}
	for _, a := range addrs {
		if ipn, ok := a.(*net.IPNet); ok {
			if ip4 := ipn.IP.To4(); ip4 != nil && strings.HasPrefix(ip4.String(), "192.168.68.") {
				return ip4.String(), nil
			}
		}
	}
	return "", fmt.Errorf("no 192.168.68.x IPv4 address on %s", name)
}

// detectNetwork finds an ethernet interface with a 192.168.68.x IPv4 address.
// Skips loopback / utun / awdl / bridge / etc.
func detectNetwork() (NetworkInfo, error) {
	ifaces, err := net.Interfaces()
	if err != nil {
		return NetworkInfo{}, err
	}
	for _, iface := range ifaces {
		if iface.Flags&net.FlagUp == 0 || iface.Flags&net.FlagLoopback != 0 {
			continue
		}
		lower := strings.ToLower(iface.Name)
		if strings.HasPrefix(lower, "lo") || strings.HasPrefix(lower, "utun") ||
			strings.HasPrefix(lower, "awdl") || strings.HasPrefix(lower, "llw") ||
			strings.HasPrefix(lower, "bridge") || strings.HasPrefix(lower, "anpi") ||
			strings.HasPrefix(lower, "ap") || strings.HasPrefix(lower, "gif") ||
			strings.HasPrefix(lower, "stf") {
			continue
		}
		addrs, err := iface.Addrs()
		if err != nil {
			continue
		}
		for _, a := range addrs {
			ipn, ok := a.(*net.IPNet)
			if !ok {
				continue
			}
			ip4 := ipn.IP.To4()
			if ip4 == nil {
				continue
			}
			if strings.HasPrefix(ip4.String(), "192.168.68.") {
				return NetworkInfo{Interface: iface.Name, IP: ip4.String()}, nil
			}
		}
	}
	return NetworkInfo{}, errors.New("no interface found on 192.168.68.0/24; plug in the operator host or pass --iface")
}

func dnsmasqAvailable() bool {
	if _, err := os.Stat("/opt/homebrew/sbin/dnsmasq"); err == nil {
		return true
	}
	_, err := exec.LookPath("dnsmasq")
	return err == nil
}

// tailReader is a no-op placeholder retained for backward compatibility.
var _ io.Reader = (*strings.Reader)(nil)
